base/strings/annotated.jl

```diff
@@ -386,3 +386,132 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
+                                         first(range):first(span)-1
+                                     else last(span)+1:last(range) end, annot)
+                # If `span` fits exactly within `range`, then we've only copied over
+                # the beginning overhang, but also need to conserve the end overhang.
+                if first(range) < first(span) && last(span) < last(range)
+                    push!(extras, (last(span)+1:last(range), annot))
+                end
+            end
+            # Insert any extra entries in the appropriate position
+            for entry in extras
+                indices = searchsorted(io.annotations, (first(entry),), by=first)
+                splice!(io.annotations, indices, Tuple{UnitRange{Int}, Pair{Symbol, Any}}[entry])
+            end
+        end
+        # Fill in the annotations from `astr`, accounting for `offset`
+        for (region, annot) in astr.annotations
+            region = first(region)+offset:last(region)+offset
+            indices = searchsorted(io.annotations, (region,), by=first)
+            splice!(io.annotations, indices, Tuple{UnitRange{Int}, Pair{Symbol, Any}}[(region, annot)])
+        end
+    else
+        # Fill in the annotations from `astr`, accounting for `offset`
+        for (region, annot) in astr.annotations
+            region = first(region)+offset:last(region)+offset
+            push!(io.annotations, (region, annot))
+        end
+    end
+    write(io.io, String(astr))
+end
+
+write(io::AnnotatedIOBuffer, c::AnnotatedChar) = write(io, AnnotatedString(c))
+write(io::AnnotatedIOBuffer, x::AbstractString) = write(io.io, x)
+write(io::AnnotatedIOBuffer, s::Union{SubString{String}, String}) = write(io.io, s)
+write(io::AnnotatedIOBuffer, b::UInt8) = write(io.io, b)
+
+function write(dest::AnnotatedIOBuffer, src::AnnotatedIOBuffer)
+    destpos = position(dest)
+    srcpos = position(src)
+    offset = destpos - srcpos
+    nb = write(dest.io, src.io)
+    # Clear out annotations that only start after `position(dest)`,
+    filter!(((region, _),) -> first(region) <= destpos, dest.annotations)
+    # Truncate any existing annotation to `position(dest)+1`
+    map!(((region, annot),) -> (first(region):min(last(region), destpos+1), annot),
+         dest.annotations, dest.annotations)
+    append!(dest.annotations,
+            ((max(1, first(region) + offset):last(region)+offset, annot)
+             for (region, annot) in src.annotations if first(region) >= srcpos))
+    nb
+end
+
+function read(io::AnnotatedIOBuffer, ::Type{AnnotatedString{T}}) where {T <: AbstractString}
+    if position(io) == 0
+        AnnotatedString(read(io.io, T), copy(io.annotations))
+    else
+        start = position(io.io)
+        annots = [(max(1, first(region) - start):last(region)-start, val)
+                  for (region, val) in io.annotations if last(region) > start]
+        AnnotatedString(read(io.io, T), annots)
+    end
```

Comment:
Can probably DRY this code a little bit:
```suggestion
    start = position(io.io)
    annotations = if start == 0
        copy(io.annotations)
    else
       typeof(io.annotations)[(max(1, first(region) - start):last(region)-start, val)
                  for (region, val) in io.annotations if last(region) > start]
    end
    return AnnotatedString(read(io.io, T), annotations)
```

Reply:
I'm not sure that this counts as DRY-er when you're only going from 44 to 42 symbols :sweat_smile:. If we want to go one half-step further, we can do `if (start = position(io)) == 0` (which I've just pushed).